x-pack/plugins/canvas/canvas_plugin_src/renderers/types.ts

```diff
@@ -0,0 +1,26 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+type GenericCallback = (callback: () => void) => void;
+
+export interface RendererHandlers {
+  done: () => void;
+  getFilter: () => string;
+  onDestroy: GenericCallback;
+  onResize: GenericCallback;
+  setFilter: (filter: string) => void;
+}
+
+export interface RendererSpec<RendererConfig = {}> {
```

Comment:
I don't know how to read this... is this saying that the `RendererSpec` is passed the `RendererConfig` object? Or maybe that it's returning `RendererConfig`? And how does that object relate to the `config` object down in the `render` function? Is it just sugar for defining that object as part of the specific spec's type?

Reply:
`RendererConfig` is a generic, which defaults to an empty object.

Note the `render` function takes a `config` object.  I'm stating that the `config` is represented by a generic, which is specified when you create a `RendererFactory`.  This then enforces the structure of the config used in the `render` method.

As a practical example, take a look at the changes to `dropdownFilter`, and the result of the autocomplete:

![Screen Shot 2019-04-03 at 3 40 23 PM](https://user-images.githubusercontent.com/297604/55511680-d2fc3200-5626-11e9-8ae7-2d8e47abab20.png)

Reply:
Yeah, I saw that as well, and I kind of understand generics, but I haven't seen a lot of them used so I'm still a little shaky when I come across them. Thanks for the breakdown, it makes a lot more sense now.